core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -0,0 +1,297 @@
+  def testEmptyCluster(): Unit = {
+    servers = makeServers(1)
+    TestUtils.waitUntilTrue(() => zkUtils.pathExists(ZkUtils.ControllerPath), "failed to elect a controller")
+    TestUtils.waitUntilTrue(() => zkUtils.readData(ZkUtils.ControllerEpochPath)._1.toInt == KafkaController.InitialControllerEpoch,
+      "broker failed to set controller epoch")
+  }
+
+  @Test
+  def testControllerEpochPersistsWhenAllBrokersDown(): Unit = {
+    servers = makeServers(1)
+    TestUtils.waitUntilTrue(() => zkUtils.pathExists(ZkUtils.ControllerPath), "failed to elect a controller")
+    TestUtils.waitUntilTrue(() => zkUtils.readData(ZkUtils.ControllerEpochPath)._1.toInt == KafkaController.InitialControllerEpoch,
+      "broker failed to set controller epoch")
+    servers.head.shutdown()
+    servers.head.awaitShutdown()
+    TestUtils.waitUntilTrue(() => !zkUtils.pathExists(ZkUtils.ControllerPath), "failed to kill controller")
+    TestUtils.waitUntilTrue(() => zkUtils.readData(ZkUtils.ControllerEpochPath)._1.toInt == KafkaController.InitialControllerEpoch,
+      "controller epoch was not persisted after broker failure")
+  }
+
+  @Test
+  def testControllerMoveIncrementsControllerEpoch(): Unit = {
+    servers = makeServers(1)
+    TestUtils.waitUntilTrue(() => zkUtils.pathExists(ZkUtils.ControllerPath), "failed to elect a controller")
+    TestUtils.waitUntilTrue(() => zkUtils.readData(ZkUtils.ControllerEpochPath)._1.toInt == KafkaController.InitialControllerEpoch,
+      "broker failed to set controller epoch")
+    servers.head.shutdown()
+    servers.head.awaitShutdown()
+    servers.head.startup()
+    TestUtils.waitUntilTrue(() => zkUtils.pathExists(ZkUtils.ControllerPath), "failed to elect a controller")
+    TestUtils.waitUntilTrue(() => zkUtils.readData(ZkUtils.ControllerEpochPath)._1.toInt == KafkaController.InitialControllerEpoch + 1,
+      "controller epoch was not persisted after broker failure")
+  }
+
+  @Test
+  def testTopicCreation(): Unit = {
+    servers = makeServers(1)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+  }
+
+  @Test
+  def testTopicCreationWithOfflineReplica(): Unit = {
+    servers = makeServers(2)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers.take(1))
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+  }
+
+  @Test
+  def testTopicPartitionExpansion(): Unit = {
```

Comment:
AddPartitionsTest also tests adding partitions, Are the partition expansion tests here overlapping with the tests there?